File: python/poo/basico/A01_Classes.py
"""
=============================================================
TEMA: Classes em Python

Nível:
🟢 Básico

Autor: Lucas Lima (ohlm1)
Última atualização: 17/07/2026
Pré-requisitos: Lógica de programação básica
Tempo estimado de leitura: 10 min
=============================================================

O QUE É?
-------------------------------------------------------------
Uma classe é o molde para criar objetos em Python. Ela define 
o conjunto de atributos (dados) e métodos (comportamentos) 
que os objetos criados a partir dela devem possuir.

-------------------------------------------------------------
COMO FUNCIONA?
-------------------------------------------------------------
Ao definir uma classe, você cria uma nova estrutura de dados. 
Quando você a chama como uma função, o Python executa o método 
`__init__` (construtor) para inicializar um novo objeto na 
memória com os estados definidos.

-------------------------------------------------------------
POR QUE EXISTE?
-------------------------------------------------------------
Sem classes, teríamos dados dispersos em variáveis simples 
ou dicionários difíceis de controlar. Classes permitem 
agrupar tudo relacionado a uma entidade em um único lugar, 
facilitando a manutenção e a reutilização.

-------------------------------------------------------------
QUANDO USAR?
-------------------------------------------------------------
Sempre que você precisar representar entidades do mundo real 
(ex: Usuário, Produto) ou gerenciar estados complexos que 
mudam ao longo da execução do programa.

-------------------------------------------------------------
QUANDO NÃO USAR?
-------------------------------------------------------------
Em scripts muito pequenos, puramente procedurais, onde o uso 
de funções simples ou dicionários resolvem o problema de 
forma mais rápida e menos complexa.

-------------------------------------------------------------
VANTAGENS
-------------------------------------------------------------
- Organização do código em blocos modulares.
- Facilita a reutilização e extensão de funcionalidades.
- Permite o uso de herança e polimorfismo.

-------------------------------------------------------------
DESVANTAGENS
-------------------------------------------------------------
- Pode adicionar complexidade (over-engineering) em scripts simples.
- Curva de aprendizado inicial para entender o uso do `self`.

-------------------------------------------------------------
ERROS MAIS COMUNS
-------------------------------------------------------------
- Esquecer de passar o `self` como primeiro argumento nos métodos.
- Tentar acessar atributos da classe sem ter criado uma instância.

-------------------------------------------------------------
BOAS PRÁTICAS
-------------------------------------------------------------
- Use nomes de classes em PascalCase.
- Mantenha métodos focados em uma única responsabilidade.

-------------------------------------------------------------
ANALOGIA
-------------------------------------------------------------
Imagine uma classe como um "molde de biscoito". O molde não 
é o biscoito, mas ele garante que todo biscoito feito com ele 
tenha o mesmo formato. Criar um objeto é o ato de pressionar 
o molde na massa.

-------------------------------------------------------------
RESUMO
-------------------------------------------------------------
Classes são moldes para objetos que unem dados e ações. Elas 
tornam o código mais organizado, reaproveitável e facilitam a 
modelagem de sistemas complexos.
-------------------------------------------------------------
"""

"""
-------------------------------------------------------------
EXEMPLO PRÁTICO
-------------------------------------------------------------
"""

"""
-------------------------------------------------------------
CÓDIGO CORRETO
-------------------------------------------------------------
"""

# Código completo (com observações)
class ContaBancaria:
    def __init__(self, titular, saldo_inicial): # self = referência ao objeto
        self.titular = titular                  # Armazena no objeto
        self.saldo = saldo_inicial              # Armazena no objeto

    def depositar(self, valor):                 # Método de instância
        self.saldo += valor                     # Atualiza atributo do objeto
        return self.saldo

    def consultar(self):                        # Acessa os dados do objeto
        return f"Titular: {self.titular} | Saldo: R$ {self.saldo:.2f}"


"""
Explicação escrita

O código correto utiliza o parâmetro 'self' como o primeiro argumento em 
todos os métodos. Isso é crucial porque o 'self' atua como uma ponte 
que conecta as variáveis (como titular e saldo) à instância específica 
da classe que foi criada. Sem o 'self', o Python não saberia diferenciar 
os dados da "Conta A" dos dados da "Conta B".
"""

"""
-------------------------------------------------------------
CÓDIGO INCORRETO
-------------------------------------------------------------
"""

# Código completo (com observações)
class ContaBancariaErrada:
    def __init__(titular, saldo_inicial): # ERRO: Faltou o 'self'
        titular = titular                 # ERRO: Apenas variável local
        saldo = saldo_inicial             # ERRO: Apenas variável local

    def depositar(valor):                 # ERRO: Faltou o 'self'
        saldo += valor                    # ERRO: 'saldo' não foi encontrado

    def consultar():                      # ERRO: Faltou o 'self'
        return f"Saldo: R$ {saldo:.2f}"   # ERRO: 'saldo' não existe no escopo


"""
Explicação escrita

Este código falha por três motivos principais: 
1. Omissão do 'self': Ao não incluir 'self' como primeiro parâmetro, o 
   Python não reconhece a relação entre a função e a instância. 
2. Escopo local: As atribuições 'titular = titular' criam variáveis 
   temporárias que são descartadas assim que o método termina; o objeto 
   continua vazio. 
3. NameError: Quando você tenta acessar 'saldo' nos métodos posteriores, 
   o interpretador não encontra essa variável, pois ela nunca foi 
   vinculada ao objeto (não foi usada com 'self.').
"""
"""
=============================================================
DESAFIO
=============================================================
Crie uma classe `Livro` com atributos `titulo` e `autor`. 
Crie um método que exibe essas informações. Explique por que 
usar uma classe é melhor do que manter o título e autor em 
variáveis soltas.

=============================================================
VOCÊ SABIA? (opcional)
=============================================================
Em Python, tudo é um objeto! Isso inclui inteiros, strings 
e até funções. Quando você define `x = 10`, você está criando 
uma instância da classe `int`.

=============================================================
REFERÊNCIAS
=============================================================
- Documentação oficial: https://docs.python.org/3/tutorial/classes.html
=============================================================
"""

File: java/poo/A01_Classes.java
package poo;

public class A01_Classes {

/*
=============================================================
TEMA: Classes e Objetos

Nível:
🟢 Básico

Autor: Morcineck
Última atualização: 13/07/2026 H01:23
Pré-requisitos: Lógica de programação básica (variáveis, tipos, métodos)
Tempo estimado de leitura: 10 min
=============================================================

O QUE É?
-------------------------------------------------------------
Uma classe é um molde, um projeto que descreve como algo deve
ser: quais características (atributos) e quais comportamentos
(métodos) esse "algo" vai ter.

Um objeto é uma instância concreta desse molde, a coisa real,
criada a partir da classe, ocupando memória e com valores
próprios.

Classe descreve. Objeto existe.

-------------------------------------------------------------
COMO FUNCIONA?
-------------------------------------------------------------
Você define a classe uma única vez, com a palavra-chave class.
A partir dela, pode criar quantos objetos quiser usando new.

Cada objeto criado tem os seus próprios valores de atributos, mesmo
compartilhando a mesma estrutura (a mesma classe). Ou seja, dois
objetos da mesma classe são "parentes" na forma, mas independentes
no conteúdo.

-------------------------------------------------------------
POR QUE EXISTE?
-------------------------------------------------------------
Sem classes, precisaríamos representar cada "coisa" do nosso
programa com variáveis soltas e sem relação organizada entre si,
por exemplo, várias variáveis separadas para nome, preço e
estoque de um produto, sem nada que amarre essas informações
como pertencentes ao mesmo produto.

Classes resolvem isso agrupando dado (atributos) e comportamento
(métodos) relacionados numa única estrutura, refletindo como
pensamos sobre o mundo real: "um Produto tem nome, preço e
estoque, e sabe calcular o seu próprio desconto."

-------------------------------------------------------------
QUANDO USAR?
-------------------------------------------------------------
Sempre que você identificar um "substantivo" importante no seu
problema, algo que tem características e comportamentos
próprios: Usuário, Pedido, Produto, Conta, Veículo.

-------------------------------------------------------------
QUANDO NÃO USAR?
-------------------------------------------------------------
Para dados soltos, temporários, sem comportamento associado (ex:
uma variável para contar um loop), criar uma classe é exagero.
Nem tudo no código precisa virar uma classe, só o que representa
um conceito com identidade própria.

-------------------------------------------------------------
VANTAGENS
-------------------------------------------------------------
- Organiza dado e comportamento relacionados num único lugar.
- Permite criar múltiplos objetos reaproveitando a mesma estrutura.
- Aproxima o código da forma como pensamos sobre o problema real.

-------------------------------------------------------------
DESVANTAGENS
-------------------------------------------------------------
- Criar classes para tudo, mesmo coisas simples, pode deixar o
  código mais verboso do que precisa ser.
- Iniciantes às vezes confundem classe com objeto, tratando os
  dois como sinônimos, o que gera confusão em conceitos futuros
  (como herança e polimorfismo).

-------------------------------------------------------------
ERROS MAIS COMUNS
-------------------------------------------------------------
- Achar que a classe "é" o objeto. A classe é só a planta; o
  objeto é a casa construída a partir dela.
- Esquecer que cada objeto tem a sua própria cópia dos atributos,
  alterar o atributo de um objeto não afeta outro objeto da
  mesma classe.
- Colocar numa única classe responsabilidades que não têm
  relação entre si (ex: uma classe "Produto" que também cuida de
  enviar e-mail).

-------------------------------------------------------------
BOAS PRÁTICAS
-------------------------------------------------------------
- Nomeie classes com substantivos, começando com letra maiúscula
  (Produto, não produto ou CriarProduto).
- Uma classe deve ter uma única responsabilidade clara.
- Prefira atributos privados (tema em Encapsulamento) em vez
  de deixá-los acessíveis livremente.

-------------------------------------------------------------
ANALOGIA
-------------------------------------------------------------
Pense numa fôrma de bolo.

A fôrma (classe) define o formato: redondo, com furo no meio.
Mas a fôrma sozinha não é um bolo, é apenas o molde.

Cada vez que você usa a fôrma, sai um bolo (objeto) diferente:
um de chocolate, outro de baunilha, cada um com o seu próprio
sabor e cobertura, mesmo tendo vindo da mesma fôrma.

A fôrma é a classe. Cada bolo assado é um objeto.

-------------------------------------------------------------
RESUMO
-------------------------------------------------------------
Classe é a definição, o molde, a estrutura, o que algo pode ter
e fazer. Objeto é a existência concreta dessa definição, criada
com new, com valores próprios. Toda a Orientação a Objetos
parte dessa distinção.
=============================================================
*/

    // Declaração de classe, aqui você define o "molde": quais atributos
    public static class Produto {
        String nome;    // <- Atributos
        double preco;  // <- Atributos
        int estoque;  // <- Atributos

    // Aqui é o construtor, onde se define como o objeto nasce, recebendo os valores dos atributos
        public Produto(String nome, double preco, int estoque) {
            this.nome = nome;
            this.preco = preco;
            this.estoque = estoque;
        }

    //  Aqui é método, o comportamento que todo objeto "sabe" fazer
    //  "void" indica que ele não devolve nenhum valor, apenas executa uma ação (imprimir os dados na tela)
        void exibirDetalhes(){
            System.out.println(nome + " - R$" + preco + " (" + estoque +" unidades)");
        }
    }

    // ❌ Errado / ingênuo: representar o mesmo conceito com
    // variáveis soltas, sem nenhuma estrutura que as agrupe.
    // Isso funciona para UM produto, mas não escala: e se você
    // precisar de 10 produtos? 100? Viraria 300 variáveis soltas,
    // sem forma de relacionar cada trio (nome/preço/estoque) entre si.
    //
    // String nomeProdutoA = "Teclado";
    // double precoProdutoA = 150.0;
    // int estoqueProdutoA = 20;


    public static void main(String[] args) {

        // Criando dois objetos a partir da mesma classe.
        // Cada um tem os seus próprios valores, eles são independentes.
        Produto produto01 = new Produto("Teclado", 150.0, 20);
        Produto produto2 = new Produto("Mouse", 80.0, 35);

        produto01.exibirDetalhes();
        produto2.exibirDetalhes();

        // Se alterar o estoque de um objeto não vai afetar o outro.
        produto01.estoque = 10;
        System.out.println("Estoque atualizado do produto01: " + produto01.estoque);
        System.out.println("Estoque do produto02 (Não alterado): " + produto2.estoque);

    }
}

/*
=============================================================
DESAFIO
=============================================================
Crie uma classe Pessoa com os atributos nome, idade e
cidade.

1. Crie um construtor que receba esses três valores.
2. Crie um método exibirDados que imprima as informações da
   pessoa (nome, idade e cidade), assim como mostrado no código a cima
   em exibirDetalhes na classe Produto.
3. No main, crie pelo menos dois objetos Pessoa diferentes, com valore
   distintos, e chame o método exibirDados de cada um.

Depois, explique com suas próprias palavras: por que faz sentido
essas informações estarem todas dentro da mesma classe, em vez
de serem variáveis soltas?


============================================================
VOCÊ SABIA?
=============================================================
Em Java, todo objeto que você cria com new vive na memória
heap, e a variável que você usa para "guardá-lo" (como
produto01) na verdade não guarda o objeto em si, ela guarda
uma referência (um "endereço") para onde o objeto está na
memória. É por isso que, se você copiar uma referência para
outra variável, as duas passam a apontar para o mesmo objeto.

=============================================================
REFERÊNCIAS
=============================================================
- Documentação oficial: https://docs.oracle.com/javase/tutorial/java/javaOO/classes.html
- Baeldung - Java Classes and Objects: https://www.baeldung.com/java-classes-objects
=============================================================
 */